hpfeeds/broker/auth/mongo.py

```diff
@@ -0,0 +1,61 @@
+import asyncio
+import logging
+
+import motor.motor_asyncio
+from hpfeeds.broker.server import ServerException
+
+# inserting a user:
+# mongo
+# db.auth_key.insert({"identifier": "testin", "secret": "secretkey", "publish": ["chan1", "chan2"], subscribe: ["chan2"]})
+# Connection string example mongodb://username:password@mongohost:27017/dbname
+
+
+class Authenticator(object):
+
+    def __init__(self, connection_string):
+        self.logger = logging.getLogger('hpfeeds.broker.auth.mongo.Authenticator')
+        self.connection = False
+        try:
+            self.dbstring, self.dbname = connection_string.rsplit('/', 1)
+        except Exception as err:
+            ServerException("Unable to create connection string: {0}".format(err))
+
+    async def db_connect(self):
+        try:
+            self.logger.debug("Connecting to Mongo DB")
+            loop = asyncio.get_event_loop()
+            self.client = motor.motor_asyncio.AsyncIOMotorClient(self.dbstring, io_loop=loop)
+            self.db = self.client[self.dbname]
+            self.collection = self.db['auth_key']
+        except Exception as err:
+            self.connection = False
+            ServerException("Unable to connect to mongo database: {0}".format(err))
+
+    async def start(self):
+        pass
+
+    async def close(self):
+        self.logger.debug("Closing Mongo Connection")
+        self.client.close()
+
+    async def get_authkey(self, ident):
+        self.logger.debug("Auth key for {0} requested".format(ident))
+
+        # We need to wait for the asyncio loop to be up
+        # So set everything up on the first auth call at which point we should be running.
+        if not self.connection:
+            await self.db_connect()
+            self.connection = True
+
+        auth_key = await self.collection.find_one({"identifier": ident})
```

Comment:
What happens if the broker loses its connection? Will it automatically reconnect? Or will this raise that the database has gone away?

Reply:
It will raise that the database has gone away. What's your preferred option?

I can put in a check and reconnect loop so it will keep trying to connect, add a counter so its not infinite looping. 

Or let it fail and raise

Reply:
What do you think about something like:

```python
try:
    auth_key = await self.collection.find_one(.....)
except Exception:
    self.connection = False
    self.logged.exception(f"Unhandled exception whilst trying to retried user {ident!r}"
    return None
```

We still get an error logged, but the connection will be retried next time AND the client will at least get an error message from the broker (if we raise the connection will just seem to hang on their side).